packages/playwright-core/src/server/recorder.ts

```diff
@@ -243,13 +245,26 @@ export class Recorder implements InstrumentationListener, IRecorder {
     this._refreshOverlay();
   }
 
+  private _scopeHighlightedSelectorToFrame(frame: Frame): string  {
```

Comment:
This fix now looks very isolated. I think we can do the right thing here, something like:

```ts
try {
  const mainFrame = frame.page().mainFrame();
  const resolved = await mainFrame.selectors.resolveFrameForSelector(this._highlightedSelector);
  // Either selector includes a FrameLocator that points to this exact frame,
  // or selector has no frame locators at all and so we highlight in all frames.
  return resolved?.frame === frame || resolved?.frame === mainFrame ? resolved.selector : '';
} catch {
  return '';
}
```

Reply:
That looks good, implemented it in https://github.com/microsoft/playwright/pull/33273/commits/217b1540fb0a1edd542197507eb243a932b7e5b8.

This is called once per frame, and they all have the same result for `resolveFrameForSelector`. We could optimise here by sharing the results, but i'm not sure it's worth the code complexity. This is only used in debugging anyways, not in CI runs.